<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<!--Converted with LaTeX2HTML 2008 (1.71)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>Before starting</TITLE>
<META NAME="description" CONTENT="Before starting">
<META NAME="keywords" CONTENT="manual">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2008">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="manual.css">

<LINK REL="next" HREF="node4.html">
<LINK REL="previous" HREF="node2.html">
<LINK REL="up" HREF="node2.html">
<LINK REL="next" HREF="node4.html">
</HEAD>

<BODY >

<DIV CLASS="navigation"><!--Navigation Panel-->
<A NAME="tex2html136"
  HREF="node4.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="../icons//next.png"></A> 
<A NAME="tex2html132"
  HREF="node2.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="../icons//up.png"></A> 
<A NAME="tex2html126"
  HREF="node2.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="../icons//prev.png"></A> 
<A NAME="tex2html134"
  HREF="node1.html">
<IMG WIDTH="40" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="../icons//contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html137"
  HREF="node4.html">About us</A>
<B> Up:</B> <A NAME="tex2html133"
  HREF="node2.html">Overview</A>
<B> Previous:</B> <A NAME="tex2html127"
  HREF="node2.html">Overview</A>
 &nbsp; <B>  <A NAME="tex2html135"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR></DIV>
<!--End of Navigation Panel-->

<H2><A NAME="SECTION00021000000000000000"></A>
<A NAME="subsection:beforestarting"></A>
<BR>
Before starting
</H2>
Over recent years, deep sequencing technology has become a powerful approach for deeply investigating small non-coding RNA (ncRNA) populations, i.e. small RNA-seq. It is now established that an increasing number of novel small ncRNA families distinct from microRNAs are generated over kingdoms from different coding/non-coding regions via various biogenesis pathways and might involve a great spectrum of biological processes. For example, two other major classes of endogenous small RNAs, Piwi-interacting RNAs (piRNAs) and endogenous small interfering RNAs (endo-siRNAs), have been identified and widely investigated in mammals [<A
 HREF="node51.html#Ghildiyal2009">7</A>]. Moreover, in other organisms like plants more classes of small ncRNA have been described indicating that a wide range of small ncRNAs exist [<A
 HREF="node51.html#Brodersen2006">3</A>].
<BR>
<BR>
However, most of the existing tools devoted to sRNA-seq analysis, are only based on miRNAs annotation and quantification, or can only be applied to one organism. Here we present a comprehensive and flexible ncRNA analysis pipeline, <SPAN  CLASS="textbf">ncPRO-seq</SPAN> (<SPAN  CLASS="textbf"><SPAN ID="txt56">N</SPAN></SPAN>on-<SPAN  CLASS="textbf"><SPAN ID="txt57">C</SPAN></SPAN>oding RNA <SPAN  CLASS="textbf"><SPAN ID="txt58">PRO</SPAN></SPAN>filing in sRNA-<SPAN  CLASS="textbf"><SPAN ID="txt59">seq</SPAN></SPAN>) (http://ncproseq.sourceforge.net), which is able to interrogate and perform detailed analysis on small RNAs derived from annotated non-coding regions in miRBase [<A
 HREF="node51.html#Kozomara2011">8</A>], Rfam [<A
 HREF="node51.html#Gardner2011">5</A>] and repeatMasker [<A
 HREF="node51.html#Smit2008">14</A>], and regions defined by users. The ncPRO-seq pipeline also has a module to identify regions significantly enriched with short reads that can not be classified as known ncRNA families. The ncPRO-seq pipeline supports input read sequences in fastq, fasta and color space format, as well as alignment results in BAM format, meaning that small RNA raw data from the 3 current major platforms (Roche-454, Illumina-Solexa and Life technologies-SOLiD) could be analyzed with this pipeline. Finally, the ncPRO-seq pipeline can be used to analyze data based on genome from metazoan to plants.

<DIV CLASS="navigation"><HR>
<!--Navigation Panel-->
<A NAME="tex2html136"
  HREF="node4.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="../icons//next.png"></A> 
<A NAME="tex2html132"
  HREF="node2.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="../icons//up.png"></A> 
<A NAME="tex2html126"
  HREF="node2.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="../icons//prev.png"></A> 
<A NAME="tex2html134"
  HREF="node1.html">
<IMG WIDTH="40" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="../icons//contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html137"
  HREF="node4.html">About us</A>
<B> Up:</B> <A NAME="tex2html133"
  HREF="node2.html">Overview</A>
<B> Previous:</B> <A NAME="tex2html127"
  HREF="node2.html">Overview</A>
 &nbsp; <B>  <A NAME="tex2html135"
  HREF="node1.html">Contents</A></B> </DIV>
<!--End of Navigation Panel-->
<ADDRESS>
Nicolas Servant
2012-05-31
</ADDRESS>
</BODY>
</HTML>
