<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<!--Converted with LaTeX2HTML 2008 (1.71)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>Genome tracks visualisation</TITLE>
<META NAME="description" CONTENT="Genome tracks visualisation">
<META NAME="keywords" CONTENT="manual">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="LaTeX2HTML v2008">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="manual.css">

<LINK REL="next" HREF="node20.html">
<LINK REL="previous" HREF="node18.html">
<LINK REL="up" HREF="node12.html">
<LINK REL="next" HREF="node20.html">
</HEAD>

<BODY >

<DIV CLASS="navigation"><!--Navigation Panel-->
<A NAME="tex2html351"
  HREF="node20.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="../icons//next.png"></A> 
<A NAME="tex2html347"
  HREF="node12.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="../icons//up.png"></A> 
<A NAME="tex2html341"
  HREF="node18.html">
<IMG WIDTH="20" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="../icons//prev.png"></A> 
<A NAME="tex2html349"
  HREF="node1.html">
<IMG WIDTH="40" HEIGHT="20" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="../icons//contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html352"
  HREF="node20.html">Search enriched regions</A>
<B> Up:</B> <A NAME="tex2html348"
  HREF="node12.html">Web interface version</A>
<B> Previous:</B> <A NAME="tex2html342"
  HREF="node18.html">User Custom gff files</A>
 &nbsp; <B>  <A NAME="tex2html350"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR></DIV>
<!--End of Navigation Panel-->

<H3><A NAME="SECTION00042700000000000000">
Genome tracks visualisation</A>
</H3>
For each sequencing library,  for each annotation family, ncPRO-seq generates two compressed track files in http://genome.ucsc.edu/goldenPath/help/bedgraph.htmlbedgraph and http://genome.ucsc.edu/FAQ/FAQformat.html#format1formatbed formats to describe read mapping in sense and antisense direction of annotation items respectively. These track files can be then, easily upload in a visualisation tool as the UCSC Genome Browser [#!Dreszer2012!#].
<BR>
Moreover, the interface allows users to select different subsets of reads mapped in the genome to generate BedGraph formatted track files.
The minimum and maximum size of the reads, as well as the minimum and maximum number of locations can be used to filter the aligned reads (Figure&nbsp;<A HREF="#fig:web6">6</A>).

<DIV ALIGN="CENTER"><A NAME="fig:web6"></A><A NAME="754"></A>
<TABLE>
<CAPTION ALIGN="BOTTOM"><STRONG>Figure 6:</STRONG>
ncPRO-seq web interface : genome tracks options</CAPTION>
<TR><TD><IMG
  WIDTH="1518" HEIGHT="231" ALIGN="BOTTOM" BORDER="0"
 SRC="./web_6.png"
 ALT="Image web_6"></TD></TR>
</TABLE>
</DIV>
<BR><HR>
<ADDRESS>
Jocelyn Brayet
2015-02-12
</ADDRESS>
</BODY>
</HTML>
